s3transfer/download.py

```diff
@@ -0,0 +1,294 @@
+                },
+                done_callbacks=[
+                    functools.partial(io_queue.put, SHUTDOWN_SENTINEL)
+                ]
+            )
+        )
+
+    def _submit_io_write_task(self, temp_filename, filename,
+                              transfer_coordinator, io_queue, done_callbacks):
+        self._io_executor.submit(
+            IOWriteTask(
+                transfer_coordinator=transfer_coordinator,
+                main_kwargs={
+                    'filename': temp_filename,
+                    'final_filename': filename,
+                    'osutil': self._osutil,
+                    'io_queue': io_queue
+                },
+                done_callbacks=done_callbacks,
+                is_final=True
+            )
+        )
+
+    def _calculate_range_param(self, part_size, part_index, num_parts):
+        # Used to calculate the Range parameter
+        start_range = part_index * part_size
+        if part_index == num_parts - 1:
+            end_range = ''
+        else:
+            end_range = start_range + part_size - 1
+        range_param = 'bytes=%s-%s' % (start_range, end_range)
+        return range_param
+
+
+class GetObjectTask(Task):
+    STREAM_CHUNK_SIZE = 8 * 1024
+
+    def _main(self, client, bucket, key, extra_args, callbacks,
+              max_attempts, io_queue, start_index=0):
+        """Downloads an object and places content into io queue
+
+        :param client: The client to use when calling GetObject
+        :param bucket: The bucket to download from
+        :param key: The key to download from
+        :param exta_args: Any extra arguements to include in GetObject request
+        :param callbacks: List of progress callbacks to invoke on download
+        :param max_attempts: The number of retries to do when downloading
+        :param io_queue: The queue that content of the object will be
+            placed in
+        :param start_index: The location in the file to start writing the
+            content of the key to.
+        """
+        last_exception = None
+        for i in range(max_attempts):
+            try:
+                response = client.get_object(
+                    Bucket=bucket, Key=key, **extra_args)
+                streaming_body = StreamReaderProgress(
+                    response['Body'], callbacks)
+
+                current_index = start_index
+                chunks = iter(
+                    lambda: streaming_body.read(self.STREAM_CHUNK_SIZE), b'')
+                for chunk in chunks:
+                    # If the io queue was shutdown just stop sending to
+                    # chunks to the queue and quit the download.
+                    try:
+                        io_queue.put((current_index, chunk))
+                    except QueueShutdownError:
+                        return
+                    current_index += len(chunk)
+                return
+            except S3_RETRYABLE_ERRORS as e:
```

Comment:
How are we handling callbacks on retry?  How would a callback handle something like displaying progress percentage?  It would have to know when retries happen right?